tests/unit/s2n_client_auth_handshake_test.c

```diff
@@ -0,0 +1,375 @@
+{
+    /* Set up client and server connections */
+    struct s2n_connection *client_conn;
+    struct s2n_connection *server_conn;
+    EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+    EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+
+    EXPECT_SUCCESS(s2n_connection_set_config(client_conn, client_config));
+    client_conn->x509_validator.skip_cert_validation = 1;
+    client_conn->server_protocol_version = S2N_TLS13;
+    client_conn->client_protocol_version = S2N_TLS13;
+    client_conn->actual_protocol_version = S2N_TLS13;
+    client_conn->secure.conn_sig_scheme = s2n_ecdsa_secp256r1_sha256;
+    client_conn->secure.client_cert_sig_scheme = s2n_ecdsa_secp256r1_sha256;
+    client_conn->secure.cipher_suite = &s2n_tls13_aes_128_gcm_sha256;
+    if (!no_cert) {
+        client_conn->handshake_params.our_chain_and_key = ecdsa_cert;
+    }
+
+    EXPECT_SUCCESS(s2n_connection_set_config(server_conn, server_config));
+    server_conn->server_protocol_version = S2N_TLS13;
+    server_conn->client_protocol_version = S2N_TLS13;
+    server_conn->actual_protocol_version = S2N_TLS13;
+    server_conn->secure.conn_sig_scheme = s2n_ecdsa_secp256r1_sha256;
+    server_conn->secure.cipher_suite = &s2n_tls13_aes_128_gcm_sha256;
+
+    if (no_cert) {
+        EXPECT_SUCCESS(s2n_connection_set_client_auth_type(server_conn, S2N_CERT_AUTH_OPTIONAL));
+        EXPECT_SUCCESS(s2n_connection_set_client_auth_type(client_conn, S2N_CERT_AUTH_OPTIONAL));
+    } else {
+        server_conn->x509_validator.skip_cert_validation = 1;
+        EXPECT_SUCCESS(s2n_connection_set_client_auth_type(server_conn, S2N_CERT_AUTH_REQUIRED));
+        EXPECT_SUCCESS(s2n_connection_set_client_auth_type(client_conn, S2N_CERT_AUTH_REQUIRED));
+    }
+
+    /* Create nonblocking pipes */
+    struct s2n_test_piped_io piped_io;
+    EXPECT_SUCCESS(s2n_piped_io_init_non_blocking(&piped_io));
+    EXPECT_SUCCESS(s2n_connections_set_piped_io(client_conn, server_conn, &piped_io));
+
+    /* Negotiate handshake */
+    EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+
+    EXPECT_TRUE(IS_CLIENT_AUTH_HANDSHAKE(server_conn->handshake.handshake_type));
+    EXPECT_TRUE(IS_CLIENT_AUTH_HANDSHAKE(client_conn->handshake.handshake_type));
+    EXPECT_EQUAL(IS_CLIENT_AUTH_NO_CERT(server_conn->handshake.handshake_type), no_cert);
+    EXPECT_EQUAL(IS_CLIENT_AUTH_NO_CERT(client_conn->handshake.handshake_type), no_cert);
+
+    const char *app_data_str = "APPLICATION_DATA";
+    EXPECT_EQUAL(strcmp(app_data_str, s2n_connection_get_last_message_name(client_conn)), 0);
+
+    /* Clean up */
+    EXPECT_SUCCESS(s2n_shutdown_test_server_and_client(server_conn, client_conn));
+    EXPECT_SUCCESS(s2n_connection_free(server_conn));
+    EXPECT_SUCCESS(s2n_connection_free(client_conn));
+    EXPECT_SUCCESS(s2n_piped_io_close(&piped_io));
+
+    return 0;
+}
+
+int s2n_test_client_auth_message_by_message(bool no_cert)
```

Comment:
What does this test provide that a standard self-talk test doesn't?

Reply:
It verifies the strict ordering of messages. Also, the way I set it up, it verifies more pre and post conditions of specific messages to ensure pieces of the connection are being updated at certain times. This actually helped me find a bug in cert_request_recv that the standard self-talk test did not catch.

Reply:
Is there any way you could just test the specific portion you care about to keep it more focused and maintainable? Like call s2n_negotiate until you hit a certain point, then do your manual handshake and special checks, then continue using s2n_negotiate again?

At a minimum, I would remove verification that's not relevant to client auth. For example, I'm guessing you don't care about any of the key generation checks currently in here, and those are left over from Josh's version of this test.

Reply:
sure, I can modify it